File: src/cowtree/projection.py
"""Project selected source paths into Git without changing a caller's index."""

from __future__ import annotations

from dataclasses import dataclass
import os
from pathlib import Path
import re
import subprocess
import tempfile

from cowtree.errors import CowtreeError, CowtreeErrorCode
from cowtree.git import GitRepository
from cowtree.install import local_path


@dataclass(frozen=True)
class ProjectionSpec:
    parent: str | None = None
    reuse: str | None = None


@dataclass(frozen=True)
class GitProjection:
    """Own source-only Git objects and refs under the cowtree namespace."""

    repository: GitRepository

    def create(
        self,
        tree: Path,
        paths: tuple[str, ...],
        parents: tuple[str, ...] = (),
        reuse: str | None = None,
    ) -> str:
        """Build a source commit with an isolated index and candidate attributes."""
        for path in paths:
            local_path(root=tree, relative=path)
        git_directory = self.repository.capture(
            args=["rev-parse", "--path-format=absolute", "--git-common-dir"]
        ).removesuffix("\n")
        with tempfile.TemporaryDirectory(prefix=".cowtree-index-", dir=tree.parent) as scratch:
            environment = os.environ.copy()
            environment["GIT_INDEX_FILE"] = str(Path(scratch) / "index")
            command = [
                "git",
                "--literal-pathspecs",
                "--git-dir",
                git_directory,
                "--work-tree",
                str(tree),
                "-c",
                "core.filemode=true",
                "-c",
                "core.fsmonitor=false",
                "-c",
                "core.sparseCheckout=false",
            ]
            self.execute(
                command=[*command, "read-tree", "--empty"], directory=tree, environment=environment
            )
            if paths:
                data = b"".join(os.fsencode(path) + b"\0" for path in paths)
                self.execute(
                    command=[
                        *command,
                        "add",
                        "--force",
                        "--pathspec-from-file=-",
                        "--pathspec-file-nul",
                    ],
                    directory=tree,
                    environment=environment,
                    data=data,
                )
            tree_id = self.execute(
                command=[*command, "write-tree"], directory=tree, environment=environment
            ).strip()
            if reuse is not None:
                self.validate_commit(commit=reuse, tree=tree_id, parents=parents)
                return reuse
            arguments = [*command, "commit-tree", tree_id]
            for parent in parents:
                if not re.fullmatch(r"[0-9a-f]{40}|[0-9a-f]{64}", parent):
                    raise CowtreeError(CowtreeErrorCode.INVALID_ARGUMENTS, "invalid Git parent")
                arguments.extend(["-p", parent])
            commit = self.execute(
                command=[*arguments, "-m", "cowtree snapshot"],
                directory=tree,
                environment=environment,
            ).strip()
        if not re.fullmatch(r"[0-9a-f]{40}|[0-9a-f]{64}", commit):
            raise CowtreeError(CowtreeErrorCode.COMMAND_FAILED, "Git returned an invalid commit")
        return commit

    def validate_commit(self, commit: str, tree: str, parents: tuple[str, ...]) -> None:
        """Reuse a checked commit only when its source tree and ordered parents match."""
        if not re.fullmatch(r"[0-9a-f]{40}|[0-9a-f]{64}", commit):
            raise CowtreeError(CowtreeErrorCode.INVALID_ARGUMENTS, "invalid validated Git identity")
        identity = self.repository.capture(
            args=["rev-parse", "--verify", "--end-of-options", f"{commit}^{{tree}}"]
        )
        ancestry = self.repository.capture(
            args=["rev-list", "--parents", "-n", "1", commit]
        ).split()
        if identity.strip() != tree or tuple(ancestry[1:]) != parents:
            raise CowtreeError(
                CowtreeErrorCode.COMMAND_FAILED, "projection differs from validated commit"
            )

    def set_ref(self, reference: str, commit: str, expected: str | None = None) -> None:
        """Publish only within refs/cowtree, with an explicit previous-value check."""
        if not reference.startswith("refs/cowtree/"):
            raise CowtreeError(CowtreeErrorCode.INVALID_ARGUMENTS, "ref is outside refs/cowtree")
        if any(
            not re.fullmatch(r"[0-9a-f]{40}|[0-9a-f]{64}", value)
            for value in (commit, expected)
            if value is not None
        ):
            raise CowtreeError(CowtreeErrorCode.INVALID_ARGUMENTS, "invalid Git ref value")
        self.repository.run(args=["check-ref-format", reference])
        current = self.repository.run(
            args=["rev-parse", "--verify", "--quiet", "--end-of-options", reference], check=False
        )
        if current.returncode == 0 and current.stdout.removesuffix("\n") == commit:
            return
        if current.returncode not in (0, 1):
            raise CowtreeError(CowtreeErrorCode.COMMAND_FAILED, current.stderr)
        previous = "0" * len(commit) if expected is None else expected
        self.repository.run(args=["update-ref", reference, commit, previous])

    def remove_ref(self, reference: str, expected: str) -> None:
        """Remove an owned ref only if it still names the expected commit."""
        if not reference.startswith("refs/cowtree/") or not re.fullmatch(
            r"[0-9a-f]{40}|[0-9a-f]{64}", expected
        ):
            raise CowtreeError(CowtreeErrorCode.INVALID_ARGUMENTS, "invalid owned Git reference")
        self.repository.run(args=["check-ref-format", reference])
        current = self.repository.run(
            args=["rev-parse", "--verify", "--quiet", "--end-of-options", reference], check=False
        )
        if current.returncode == 1:
            return
        if current.returncode != 0:
            raise CowtreeError(CowtreeErrorCode.COMMAND_FAILED, current.stderr)
        self.repository.run(args=["update-ref", "-d", reference, expected])

    def execute(
        self,
        command: list[str],
        directory: Path,
        environment: dict[str, str],
        data: bytes | None = None,
    ) -> str:
        result = subprocess.run(  # noqa: S603
            command,
            cwd=directory,
            env=environment,
            input=data,
            capture_output=True,
            check=False,
            pass_fds=self.repository.io.descriptors,
        )
        if result.returncode != 0:
            raise CowtreeError(
                CowtreeErrorCode.COMMAND_FAILED, result.stderr.decode(errors="replace")
            )
        output = result.stdout.decode("utf-8")
        return output

File: crates/workspace/src/projection.rs
// Copyright (c) 2026 Windsor Nguyen

//! Build source-only Git commits with private indexes and compare-and-swap refs.

use std::{path::Path, process::Command};

use cowtree_metadata::Snapshot;

use crate::{CommitId, Error, Result, error::Issue, git::Repository, paths};

impl Repository {
    pub(crate) fn project(
        &self,
        tree: &Path,
        manifest: &Snapshot,
        parents: &[CommitId],
        reuse: Option<&CommitId>,
    ) -> Result<CommitId> {
        let directory = tempfile::Builder::new()
            .prefix(".cowtree-index-")
            .tempdir_in(tree.parent().ok_or_else(|| Issue::InvalidRoot.at(tree))?)
            .map_err(|error| Error::io(tree, error))?;
        let common = &self.directory;
        let command = || -> Result<Command> {
            let mut command = self.command()?;
            command
                .args(["--literal-pathspecs", "--git-dir"])
                .arg(common)
                .arg("--work-tree")
                .arg(tree)
                .args([
                    "-c",
                    "core.filemode=true",
                    "-c",
                    "core.fsmonitor=false",
                    "-c",
                    "core.sparseCheckout=false",
                ])
                .env("GIT_INDEX_FILE", directory.path().join("index"));
            Ok(command)
        };
        if !manifest.is_empty() {
            let mut input = Vec::new();
            for path in manifest.keys() {
                paths::local(tree, path.as_str())?;
                input.extend_from_slice(path.as_str().as_bytes());
                input.push(0);
            }
            self.input(
                command()?.args([
                    "add",
                    "--force",
                    "--pathspec-from-file=-",
                    "--pathspec-file-nul",
                ]),
                &input,
            )?;
        }
        let tree_id = self.text(command()?.arg("write-tree"))?;
        if let Some(commit) = reuse {
            self.verify_projection(tree, commit, &tree_id, parents)?;
            return Ok(commit.clone());
        }
        let mut commit = command()?;
        commit.args(["commit-tree", &tree_id]);
        for parent in parents {
            commit.args(["-p", parent.as_str()]);
        }
        CommitId::parse(self.text(commit.args(["-m", "cowtree snapshot"]))?)
    }

    fn verify_projection(
        &self,
        tree: &Path,
        commit: &CommitId,
        tree_id: &str,
        parents: &[CommitId],
    ) -> Result<()> {
        let record = self.text(self.command()?.args([
            "show",
            "--no-patch",
            "--no-show-signature",
            "--format=%T%x00%P",
            commit.as_str(),
            "--",
        ]))?;
        let (identity, ancestry) =
            record.split_once('\0').ok_or_else(|| Issue::IncompleteRecord.at(tree))?;
        let expected: Vec<_> = parents.iter().map(CommitId::as_str).collect();
        if identity != tree_id || ancestry.split_whitespace().collect::<Vec<_>>() != expected {
            return Err(Issue::NodeChanged.at(tree));
        }
        Ok(())
    }

    pub(crate) fn set_ref(
        &self,
        reference: &str,
        commit: &CommitId,
        expected: Option<&CommitId>,
    ) -> Result<()> {
        self.check_ref(reference)?;
        let current = self.output(self.command()?.args([
            "rev-parse",
            "--verify",
            "--quiet",
            "--end-of-options",
            reference,
        ]))?;
        if current.status.success() && current.stdout == format!("{}\n", commit.as_str()).as_bytes()
        {
            return Ok(());
        }
        if !matches!(current.status.code(), Some(0 | 1)) {
            return Err(cowtree_git::Error::Command {
                status: current.status,
                stderr: String::from_utf8_lossy(&current.stderr).into_owned(),
            }
            .into());
        }
        let absent = "0".repeat(commit.as_str().len());
        let previous = expected.map_or(absent.as_str(), CommitId::as_str);
        self.capture(self.command()?.args(["update-ref", reference, commit.as_str(), previous]))?;
        Ok(())
    }

    pub(crate) fn remove_ref(&self, reference: &str, expected: &CommitId) -> Result<()> {
        self.check_ref(reference)?;
        let current = self.output(self.command()?.args([
            "rev-parse",
            "--verify",
            "--quiet",
            "--end-of-options",
            reference,
        ]))?;
        if current.status.code() == Some(1) {
            return Ok(());
        }
        if !current.status.success() {
            return Err(cowtree_git::Error::Command {
                status: current.status,
                stderr: String::from_utf8_lossy(&current.stderr).into_owned(),
            }
            .into());
        }
        self.capture(self.command()?.args(["update-ref", "-d", reference, expected.as_str()]))?;
        Ok(())
    }

    fn check_ref(&self, reference: &str) -> Result<()> {
        if !reference.starts_with("refs/cowtree/") {
            return Err(Issue::OutsideNamespace.at(Path::new(reference)));
        }
        self.capture(self.command()?.args(["check-ref-format", reference]))?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use cowtree_metadata::{Entry, EntryKind, ResourcePath, objects::ObjectId};

    #[test]
    fn projections_isolate_the_index_and_verify_reused_tree_and_ancestry()
    -> std::result::Result<(), Box<dyn std::error::Error>> {
        let directory = tempfile::tempdir()?;
        let root = directory.path();
        let repository = cowtree_git::Client::at(root);
        for args in [
            vec!["init", "-q"],
            vec!["config", "user.name", "Test"],
            vec!["config", "user.email", "test@example.invalid"],
        ] {
            repository.capture(repository.command()?.args(args))?;
        }
        let repository = Repository::discover(root)?;
        std::fs::write(root.join("selected file"), b"selected")?;
        std::fs::write(root.join("unselected"), b"private")?;
        repository.capture(repository.command()?.args(["add", "unselected"]))?;
        let before = repository.capture(repository.command()?.args(["ls-files", "--stage"]))?;
        let selected = Snapshot::from([(
            ResourcePath::parse("selected file")?,
            Entry { object: ObjectId::from_bytes(b"selected"), kind: EntryKind::File },
        )]);
        let initial = repository.project(root, &selected, &[], None)?;
        assert_eq!(
            repository.text(repository.command()?.args([
                "ls-tree",
                "--name-only",
                initial.as_str()
            ]))?,
            "selected file"
        );
        let empty = repository.project(root, &Snapshot::new(), &[], None)?;
        assert!(
            repository
                .capture(repository.command()?.args(["ls-tree", "--name-only", empty.as_str()]))?
                .is_empty()
        );
        let parents = [initial.clone(), empty.clone()];
        let merged = repository.project(root, &selected, &parents, None)?;
        assert_eq!(repository.project(root, &selected, &parents, Some(&merged))?, merged);
        assert!(repository.project(root, &selected, &[empty, initial], Some(&merged)).is_err());
        assert!(repository.project(root, &Snapshot::new(), &parents, Some(&merged)).is_err());
        assert_eq!(
            repository.capture(repository.command()?.args(["ls-files", "--stage"]))?,
            before
        );
        Ok(())
    }
}
